File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.*;

class Solution {
    public long[] solution(long k, long[] room_number) {
        long[] answer = new long[room_number.length];
        
        HashMap<Long, Long> checkDup = new HashMap<>();
        
        for (int i = 0; i < room_number.length; i++) {
            long r = room_number[i];
            if (checkDup.containsKey(r)) {
                List<Long> tmp = new LinkedList<>();
                tmp.add(r);
                
                Long next = checkDup.get(r);
                while (checkDup.containsKey(next)) {
                    tmp.add(next);
                    next = checkDup.get(next);
                }

                for (long t : tmp)
                    checkDup.put(t, next + 1);
                    
                checkDup.put(next, next + 1);
                answer[i] = next;
            }  
            
            else {
                checkDup.put(r, r + 1);
                answer[i] = r;
            }
        }
        return answer;
    }
}

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

unordered_map<long long, long long> checkDup;

vector<long long> solution(long long k, vector<long long> room_number) {
    vector<long long> answer;
    
    for (long long n : room_number) {
        if (checkDup.find(n) == checkDup.end()) {
            checkDup[n] = n + 1;
            answer.push_back(n);
        }
        else {
            vector<long long> tmp;
            
            tmp.push_back(n);
            long long nextRoomNumber = checkDup[n];
            
            while (checkDup.find(nextRoomNumber) != checkDup.end()) {
                tmp.push_back(nextRoomNumber);
                nextRoomNumber = checkDup[nextRoomNumber];
            }
            for (long long next : tmp)
                checkDup[next] = nextRoomNumber + 1;
            
            checkDup[nextRoomNumber] = nextRoomNumber + 1;
            answer.push_back(nextRoomNumber);
        }
    }
    return answer;
}